examples/dqn/dqn_atari.py

```diff
@@ -0,0 +1,338 @@
+            "train/q_values",
+            (data["action_value"] * data["action"]).sum().item() / frames_per_batch,
+            collected_frames,
+        )
+        episode_rewards = data["next", "episode_reward"][data["next", "done"]]
+        if len(episode_rewards) > 0:
+            episode_length = data["next", "step_count"][data["next", "done"]]
+            logger.log_scalar(
+                "train/reward", episode_rewards.mean().item(), collected_frames
+            )
+            logger.log_scalar(
+                "train/episode_length",
+                episode_length.sum().item() / len(episode_length),
+                collected_frames,
+            )
+
+        # optimization steps
+        training_start = time.time()
+        q_losses = TensorDict({}, batch_size=[cfg.loss.num_updates])
+        for j in range(cfg.loss.num_updates):
+
+            sampled_tensordict = replay_buffer.sample()
+
+            loss_td = loss_module(sampled_tensordict)
+            q_loss = loss_td["loss"]
+            optimizer.zero_grad()
+            q_loss.backward()
+            # grad_norm = torch.nn.utils.clip_grad_norm_(list(loss_module.parameters()), max_norm=0.5)
+            optimizer.step()
+            target_net_updater.step()
+            q_losses[j] = loss_td.select("loss").detach()
+
+        # Log training losses, epsilon and times
+        training_time = time.time() - training_start
+        q_losses = q_losses.apply(lambda x: x.float().mean(), batch_size=[])
+        for key, value in q_losses.items():
+            logger.log_scalar("train/" + key, value.item(), collected_frames)
+        logger.log_scalar("train/epsilon", model_explore.eps, collected_frames)
+        logger.log_scalar("train/sampling_time", sampling_time, collected_frames)
+        logger.log_scalar("train/training_time", training_time, collected_frames)
+
+        with torch.no_grad(), set_exploration_type(ExplorationType.MODE):
+            if (collected_frames - frames_per_batch) // cfg.logger.test_interval < (
+                collected_frames // cfg.logger.test_interval
+            ):
+                model.eval()
+                test_rewards = []
+                for _ in range(cfg.logger.num_test_episodes):
+                    td_test = test_env.rollout(
+                        policy=model,
+                        auto_reset=True,
+                        auto_cast_to_device=True,
+                        break_when_any_done=True,
+                        max_steps=10_000_000,
+                    )
+                    reward = td_test["next", "episode_reward"][td_test["next", "done"]]
+                    test_rewards = np.append(test_rewards, reward.cpu().numpy())
+                    del td_test
+                logger.log_scalar("eval/reward", test_rewards.mean(), collected_frames)
+                model.train()
+
+        # update weights of the inference policy
+        collector.update_policy_weights_()
+        sampling_start = time.time()
```

Comment:
To keep the training loop short, we can put the testing in a separate function, as well as the calls to the logger